linesearch.go

```diff
@@ -35,96 +39,149 @@ func (ls *LinesearchMethod) Init(loc *Location, xNext []float64) (EvaluationType
 	dim := len(loc.X)
 	ls.x = resize(ls.x, dim)
 	ls.dir = resize(ls.dir, dim)
+
 	ls.first = true
+	ls.nextMajor = false
+
+	copyLocation(&ls.loc, loc)
+	// Indicate that all fields of ls.loc are valid.
+	ls.eval = FuncEvaluation | GradEvaluation
+	if ls.loc.Hessian != nil {
+		ls.eval |= HessEvaluation
+	}
+
+	ls.lastStep = math.NaN()
+	ls.lastOp = NoOperation
 
-	return ls.initNextLinesearch(loc, xNext)
+	return ls.initNextLinesearch(loc.X)
 }
 
-func (ls *LinesearchMethod) Iterate(loc *Location, xNext []float64) (EvaluationType, IterationType, error) {
-	if ls.iterType == SubIteration {
-		// We needed to evaluate invalid fields of Location. Now we have them
-		// and can announce MajorIteration.
-		copy(xNext, loc.X)
-		ls.evalType = NoEvaluation
-		ls.iterType = MajorIteration
-		return ls.evalType, ls.iterType, nil
-	}
+func (ls *LinesearchMethod) Iterate(loc *Location) (Operation, error) {
+	switch ls.lastOp {
+	case NoOperation:
+		// TODO(vladimir-ch): We have previously returned with an error and
+		// Init() was not called. What to do? What about ls's internal state?
```

Comment:
Yes, please. The tests are passing now, and we'll be changing anyway with
the copying issue. This way we can think of the algorithm on the whole
without am the extra changes
On Sep 23, 2015 6:16 PM, "Vladimír Chalupecký" <notifications@github.com>
wrote:

> In linesearch.go
> <https://github.com/gonum/optimize/pull/123#discussion_r40274267>:
>
> >  }
> >
> > -func (ls *LinesearchMethod) Iterate(loc *Location, xNext []float64) (EvaluationType, IterationType, error) {
> > -	if ls.iterType == SubIteration {
> > -		// We needed to evaluate invalid fields of Location. Now we have them
> > -		// and can announce MajorIteration.
> > -		copy(xNext, loc.X)
> > -		ls.evalType = NoEvaluation
> > -		ls.iterType = MajorIteration
> > -		return ls.evalType, ls.iterType, nil
> > -	}
> > +func (ls *LinesearchMethod) Iterate(loc *Location) (Operation, error) {
> > +	switch ls.lastOp {
> > +	case NoOperation:
> > +		// TODO(vladimir-ch): We have previously returned with an error and
> > +		// Init() was not called. What to do? What about ls's internal state?
>
> initNextLinesearch() then overwrites lastOp with the operation returned
> from Linesearcher (which must not return NoOperation) so theoretically we
> could, but probably we should not. We will still have to decide what is the
> proper thing to do after an error return. I will modify the TODO text and
> file an issue so that we don't forget.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gonum/optimize/pull/123/files#r40274267>.
>